modules/apps/static/portal-k8s-agent/portal-k8s-agent-impl/src/main/java/com/liferay/portal/k8s/agent/internal/ClientExtensionConfigBundleTracker.java

```diff
@@ -0,0 +1,312 @@
+						"(.cx.config.key=", virtualInstancePid, ")"));
+
+			if (ArrayUtil.isNotEmpty(configurations)) {
+				configuration = configurations[0];
+
+				Dictionary<String, Object> configurationProperties =
+					configuration.getProperties();
+
+				if (Objects.equals(
+						configurationProperties.get(
+							".cx.config.bundle.last.modified"),
+						bundle.getLastModified())) {
+
+					if (_log.isInfoEnabled()) {
+						_log.info(
+							"Configuration and Bundle resource versions are " +
+								"identical");
+					}
+
+					return configuration;
+				}
+			}
+			else {
+				configuration = ConfigurationUtil.getConfiguration(
+					_configurationAdmin, virtualInstancePid);
+			}
+
+			Set<Configuration.ConfigurationAttribute> configurationAttributes =
+				configuration.getAttributes();
+
+			if (configurationAttributes.contains(
+					Configuration.ConfigurationAttribute.READ_ONLY)) {
+
+				configuration.removeAttributes(
+					Configuration.ConfigurationAttribute.READ_ONLY);
+			}
+
+			properties.put(".cx.config.bundle.id", bundle.getBundleId());
+			properties.put(
+				".cx.config.bundle.last.modified", bundle.getLastModified());
+			properties.put(".cx.config.key", virtualInstancePid);
+
+			if (_log.isInfoEnabled()) {
+				_log.info("Processed configuration " + properties);
+			}
+
+			configuration.updateIfDifferent(properties);
+
+			configuration.addAttributes(
+				Configuration.ConfigurationAttribute.READ_ONLY);
+
+			return configuration;
+		}
+		finally {
+			InMemoryOnlyConfigurationThreadLocal.setInMemoryOnly(false);
+		}
+	}
+
+	private List<Configuration> _processConfigurations(
+			Bundle bundle, String fileName, String json)
+		throws Exception {
+
+		JSONUtil.Report report = new JSONUtil.Report();
+
+		ConfigurationFile configurationFile =
+			ConfigurationUtil.getConfigurationFile(
+				bundle, fileName, json, report);
+
+		for (String error : report.errors) {
+			_log.error(error);
+		}
+
+		for (String warning : report.warnings) {
+			if (_log.isWarnEnabled()) {
+				_log.warn(warning);
+			}
+		}
+
+		if (configurationFile == null) {
+			return null;
+		}
+
+		List<Configuration> configurations = new ArrayList<>();
+
+		for (Config config : configurationFile.getConfigurations()) {
+			try {
+				configurations.add(_processConfiguration(bundle, config));
+			}
+			catch (Exception exception) {
+				_log.error(
+					"Unable to process configuration " + config, exception);
+			}
+		}
+
+		return configurations;
+	}
+
+	private static final Log _log = LogFactoryUtil.getLog(
+		ClientExtensionConfigBundleTracker.class);
+
+	private BundleTracker<List<Configuration>> _bundleTracker;
+
+	@Reference
+	private ConfigurationAdmin _configurationAdmin;
+
+	private class ClientExtensionConfigBundleTrackerCustomizer
+		implements BundleTrackerCustomizer<List<Configuration>> {
+
+		@Override
+		public List<Configuration> addingBundle(
+			Bundle bundle, BundleEvent bundleEvent) {
+
+			if (DBUpgrader.isUpgradeClient()) {
```

Comment:
@gamerson in case starting from upgrade client, you should avoid creating/opening the BundleTracker completely, rather than muting on the tracking path.